c10/core/Device.cpp

```diff
@@ -38,49 +53,24 @@ DeviceType parse_type(const std::string& device_string) {
 }
 } // namespace
 
-// `std::regex` is still in a very incomplete state in GCC 4.8.x,
-// so we have to do our own parsing, like peasants.
-// https://stackoverflow.com/questions/12530406/is-gcc-4-8-or-earlier-buggy-about-regular-expressions
-//
-// Replace with the following code once we shed our GCC skin:
-//
-// static const std::regex regex(
-//     "(cuda|cpu)|(cuda|cpu):([0-9]+)|([0-9]+)",
-//     std::regex_constants::basic);
-// std::smatch match;
-// const bool ok = std::regex_match(device_string, match, regex);
-// TORCH_CHECK(ok, "Invalid device string: '", device_string, "'");
-// if (match[1].matched) {
-//   type_ = parse_type_from_string(match[1].str());
-// } else {
-//   if (match[2].matched) {
-//     type_ = parse_type_from_string(match[1].str());
-//   } else {
-//     type_ = Type::CUDA;
-//   }
-//   AT_ASSERT(match[3].matched);
-//   index_ = std::stoi(match[3].str());
-// }
 Device::Device(const std::string& device_string) : Device(Type::CPU) {
   TORCH_CHECK(!device_string.empty(), "Device string must not be empty");
-  auto index = device_string.find(':');
-  if (index == std::string::npos) {
-    type_ = parse_type(device_string);
-  } else {
-    std::string s;
-    s = device_string.substr(0, index);
-    TORCH_CHECK(!s.empty(), "Device string must not be empty");
-    type_ = parse_type(s);
 
-    std::string device_index = device_string.substr(index + 1);
+  // We assume gcc 5+, so we can use proper regex.
+  static const std::regex regex("([a-zA-Z_]+)(?::([1-9]\\d*|0))?");
```

Comment:
Can you add a test to check that some larger number like `cuda:20` works? (the \\d looks a little weird but I think it is right and want to be sure)

Reply:
done!